### PR TITLE
Check RSA public key strength before signing

### PR DESCRIPTION
**Pull request by Stype**
Don't sign public keys when,
 * public key is less than 2048 bits
 * key exponent is less than (2^16)+1
 * modulus is even
 * modulus is divisible by known small primes

These are essentially the same checks done by the Let's Encrypt CA.

### CODE REVIEW

File: bless/ssh/public_keys/rsa_public_key.py [hunk 1]
```diff
@@ -12,6 +12,34 @@
...
+        293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359,
+        367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431,
+        433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491,
+        499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571,
+        577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641,
+        643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709,
+        719, 727, 733, 739, 743, 751
+    ]
+    for prime in small_primes:
+        if prime >= n:
```
Comment (russell-lewis) [inline bless/ssh/public_keys/rsa_public_key.py]: What is the intent behind 36-37?  This doesn't seem like the intended behavior of this function.
It looks like line 81 should prevent this check from ever being true.

**Comment by russell-lewis** [comment]
Adding these extra public key checks makes sense.  Is there a reason you didn't add validate_for_signing() checks into the RSAPublicKey constructor?  
I think it would be a good fit in there, in addition to the checking done by cryptography.load_ssh_public_key().

File: bless/ssh/public_keys/rsa_public_key.py (hunk 1 above)
Comment (Stype) [inline bless/ssh/public_keys/rsa_public_key.py]: Yeah, that was redundant

**Comment by Stype** [comment]
> Adding these extra public key checks makes sense. Is there a reason you didn't add validate_for_signing() checks into the RSAPublicKey constructor?

It probably would have made more sense if I included my change to get_ssh_public_key(). Moved the repo around right before submission and missed that one..

My thinking was that if someone needed to create a weak RSAPublicKey object for some reason, they could still do that without accidentally removing the validation checks in a patch. But it should be easy to see that the validation is called for all public key types when the lambda runs. So probably calling it from one of the factory methods. It would probably be better to have all SSHPublicKey objects to have a validate_for_signing method that each type would implement, and then have the certificate builder factory call it, but that might be overkill if no one is planning to create weak SSHPublicKey objects.

I'm happy to just call it in the constructor if you want to keep things simple.

**Pull request merged**